src/base/detail/CollectionImpl.hh

```diff
@@ -169,18 +179,23 @@ struct CollectionAssigner<Ownership::value, MemSpace::device>
     }
 };
 
-//---------------------------------------------------------------------------//
-//! Check that sizes are acceptable when taking references
-template<>
-struct CollectionStorageValidator<Ownership::value>
+#else
+// Give a useful error message when trying to use "value" collections when
+// building device code
+template<class T, MemSpace M>
+struct CollectionStorage<T, Ownership::value, M>
 {
-    template<class Size, class OtherSize>
-    void operator()(Size dst, OtherSize src)
-    {
-        CELER_VALIDATE(dst == src,
-                       << "collection is too large (" << sizeof(Size)
-                       << "-byte int cannot hold " << src << " elements)");
-    }
+    static_assert(sizeof(T) == 0,
```

Comment:
I assume this "clearer" spelling does not work:
```suggestion
    static_assert(false,
```

Reply:
Indeed not; in fact the CI is showing that the `static_assert(false` I added below needs to depend on the template parameters. This is because the compiler always evaluates `static_assert` if it can or defers if it can't. In the `false` case for a fully defined class, it's always evaluated.